Docs: Update text input usage (#3115)
carbon-design-system/carbon-website: src/pages/components/text-input/usage.mdx

<AnchorLink>Live demo</AnchorLink>
<AnchorLink>Formatting</AnchorLink>
<AnchorLink>Content</AnchorLink>
<AnchorLink>Universal behaviors</AnchorLink>
<AnchorLink>Text input</AnchorLink>
<AnchorLink>Text area</AnchorLink>
<AnchorLink>References</AnchorLink>
<AnchorLink>Feedback</AnchorLink>

</AnchorLinks>

## Overview

Text inputs enable users to enter free-form text data. The type of text field
used should reflect the length of the content you expect the user to enter. The
default text input is for short, one-line content, whereas text area is for
longer, multi-line entries.

#### When to use

- A user needs to input unique information that cannot be predicted with a
  preset of options.
- A user needs to input memorable data that can be entered more quickly in a
  free-hand format versus a more complex control.

#### When not to use

> - If there is a predefined or limited set of valid options a use can enter that

Author: No, it should be "can" but I can see how that's confusing as a "not to use" example. How about something like:

"If a user can only enter a limited set of valid options and a free-from entry will likely result in an error, then consider using a different selection control such as a dropdown, select, or radio button group."

> - If there is a predefined or limited set of valid options a use can enter that

Reviewer: Oh I see, yes that sounds better!

| ----------- | --------------- | --------------------------------------------------------------------------------------------------------------------------------------------------- |
| Small (sm)  | 32 / 2          | Use when space is constricted or when placing a text input in a form that is long and complex.                                                      |
| Medium (md) | 40 / 2.5        | This is the default size and the most commonly used size. When in doubt, use the medium size.                                                       |
| Large (lg)  | 48/3            | Use when there is a lot of space to work with. The large size is typically used in simple forms or when a text input is placed by itself on a page. |

<Row>
<Column colLg={8}>

![Text input heights](images/text-input-usage-size.png)

</Column>
</Row>

#### Text area height

Text area has a variable height that can be lengthened or shorten by the user
using the `resize` handle in the bottom right of the field. By default, text
area has a minimum height of 40px/2.5rem but no maximum height.

<Row>
<Column colLg={8}>

> ![Text input heights](images/text-area-usage-height.gif)

Author: ugh I need to re-record it, it needs to be at 2x

> ![Text input heights](images/text-area-usage-height.gif)

Author: Ok I couldn't figure out how to make it 2x in gif recording so I just made it an image example instead.

> ![Text input heights](images/text-area-usage-height.gif)

Reviewer: Ok cool!

  invalid state temporarily replaces the helper text.

#### Placeholder text

Optional placeholder text provides hints or examples of what to enter.
Placeholder text disappears after the user begins entering data into the input
and should not contain crucial information.

- Use sentence-style capitalization, and in most cases, write the text as a
  direct statement without punctuation.
- Placeholder text is not required and by default not shown in text input
  fields.
- Placeholder text can be harmful to user interactions and should only be added
  when necessary.

#### Accessibility best practices

- Labels must be announced to the screen reader on focus.
- Ensure the helper text that appears under an input is read when an assistive
  technology user stops at an input using ARIA.

### Overflow content

#### Overflow in a text input

If a user input is unexpectedly too long for the single line of a text input
then the value content can horizontally scroll inside the field container when
moving the cursor from one end of the value to the other.

#### Overflow in a text area

If a user's context exceeds the vertical space of the variable text area then a

> user can either expand the field container using the sizer control or they can

Reviewer: ```suggestion
user can either expand the field container using the resize handle or they can
```

| _Disabled_  | When the user is not allowed to interact with the text input due to either permissions, dependencies, or pre-requisites. The disabled state completely removes the interactive function from a component. The styling is not subject to WCAG contrast compliance. |
| _Skeleton_  | Use on an initial page load to indicate that the text input has not yet fully loaded.                                                                                                                                                                             |
| _Read-only_ | _Coming soon!_                                                                                                                                                                                                                                                    |

<Row>
<Column colLg={12}>

![Text input states](images/text-input-states.png)

</Column>
</Row>

### Validation

Real-time validation helps streamline the process and keep data clean when the
user fills out forms. Otherwise, validate the text input data when the user
submits the associated form. For complete guidance on error and validation, see

> the form pattern.

Reviewer: ```suggestion
the [form pattern](https://carbondesignsystem.com/patterns/forms-pattern/).
```
Suggestion to link out to the form pattern here to make it easier for people to find.

#### Invalid

An error state is triggered if the data is invalid or a required field is left
empty. Error states have three visual indicators to signify invalid content: a
red border, an error icon indicator, and an error message.

<Row>
<Column colLg={8}>

![Example of an error state being triggered](images/text-input-usage-2.gif)

</Column>
</Row>

### Default values

Where possible, add programmatic assistance. Detect and pre-fill inputs to
reduce errors and save time. When the software can’t determine the value that
belongs in an input, use type-ahead to make suggestions. Use sentence-case for
default values, detected values, and auto-completion text.

### Required vs. optional

Text inputs can be labeled as either optional or required depending on the
depending on the circumstance. For more guidance on requirement labeling, see

> the form pattern.

Reviewer: ```suggestion
the [form pattern](https://carbondesignsystem.com/patterns/forms-pattern/).
```
Same with linking out to the form pattern here

<DoDontRow>
<DoDont
colMd={4} colLg={4}
caption="If the majority of the fields are required, mark only the optional field labels with (optional).">

> ![Example showing optional field indicator](./images/text-input-usage-optional-do.png)

Reviewer: - These do/don't images need a side border because they bleed into the page.
- Also it seems like the arrow in the right image is not green.

<img width="794" alt="Screen Shot 2022-08-26 at 3 07 47 PM" src="https://user-images.githubusercontent.com/43969356/186974690-7daf8a7e-d4bb-48fd-af4e-84b264ba5046.png">
.

| Key                                                            | Interaction                                                                            |
| -------------------------------------------------------------- | -------------------------------------------------------------------------------------- |
| _Tab_                                                          | Brings focus to the text input.                                                        |
| _Enter_ or _Space_                                             | Opens any associated actions added to the input, such as a password visibility toggle. |
| _Esc_                                                          | Closes any associated actions added to the input.                                      |
| _ctrl_ + _left/right arrows_ <br/> _opt_ + _left/right arrows_ | Moves you word by word inside the field.                                               |
| _ctrl_ + _up/down arrow_ <br/> _opt_ +_up/down arrows_         | Relocates you to the start or end of the input content.                                |

## Text input

Use a text input when the expected user input is a single line of text. Text

> inputs have a fixed height and are used a simple free-form data entry. Users can

Reviewer: ```suggestion
inputs have a fixed height and are used as a simple free-form data entry. Users can
```

| _ctrl_ + _left/right arrows_ <br/> _opt_ + _left/right arrows_ | Moves you word by word inside the field.                                               |
| _ctrl_ + _up/down arrow_ <br/> _opt_ +_up/down arrows_         | Relocates you to the start or end of the input content.                                |

Text fields and text areas may have a character limit and counter. Use the
character limit and counter to let users know how long their entry can be before
they begin typing.
## Text input

While useful for data collection, character count is not an intuitive way to
communicate content length to users. Character count allows a user to understand
an input error if they surpass the character limit.
Use a text input when the expected user input is a single line of text. Text
inputs have a fixed height and are used a simple free-form data entry. Users can
enter any combination of letters, numbers, or symbols.

<Row>
<Column colLg={8}>

> ![An example of a text area.](images/text-area-usage-1.png)

Reviewer: - This images needs side borders, too.

<img width="783" alt="Screen Shot 2022-08-26 at 3 17 42 PM" src="https://user-images.githubusercontent.com/43969356/186976234-ca11d1fe-ee9e-4c92-bf31-551b38eba119.png">
.

> Password input is a sub-variant of text-input. It is used to collect private

Reviewer: ```suggestion
Password input is a sub-variant of text input. It is used to collect private
```

> character allowed or date structure.

Reviewer: ```suggestion
characters allowed or date structure.
```

<Row>
<Column colLg={8}>

![Example of a password input](images/text-input-usage-password.png)

</Column>
</Row>

## Text area

Use a text area when the expected user input is more than a few words and could
span multiple lines. It is commonly used for features like user commentary or
descriptions. It supports all the same states and functionality as text input
except for the password functionality. Text area has several unique
functionalities not included in the default text input, like the resize handle
and character counter.

<Row>
<Column colLg={8}>

![Example of a text area](images/text-input-usage-2.png)

</Column>
</Row>

### Resize handle

> Included by default in text area is the `resize` handle. It allows a user the

Reviewer: ```suggestion
Included by default in text area is the `resize` handle. It allows a user to
```

manipulate the field height by making it longer or shorter. The resize handle
has no effect on the width of the text area container, it only effects the
height. If the user makes the field size shorter than the content inside the
field then a vertical scroll will become available.

<Row>
<Column colLg={8}>

![Example showing the resize handle in text area](images/text-area-usage-resize.png)

</Column>
</Row>

### Character counter

A character counter can be added to text area to indicate both the number of
characters being entered and the total number of characters allowed. Once the

> max number of characters is reached the text area should should prevent the user

Reviewer: ```suggestion
max number of characters is reached the text area should prevent the user
```

manipulate the field height by making it longer or shorter. The resize handle
has no effect on the width of the text area container, it only effects the
height. If the user makes the field size shorter than the content inside the
field then a vertical scroll will become available.

<Row>
<Column colLg={8}>

![Placeholder text example](images/text-input-usage-1.png)
![Example showing the resize handle in text area](images/text-area-usage-resize.png)

</Column>
</Row>

### Helper text
### Character counter

Helper text is pertinent information that assists the user in completing a
field. Helper text is always available when the input is focused and appears
underneath the label. Use sentence-style capitalization, and in most cases,
write the text as full sentences with punctuation.
A character counter can be added to text area to indicate both the number of
characters being entered and the total number of characters allowed. Once the
max number of characters is reached the text area should should prevent the user
from entering any additional character and provide messaging to the user that a
limit has been met.

<Row>
<Column colLg={8}>

> ![Helper text example](images/text-input-usage-3.png)

Reviewer: - This image has a typo. It should be `charged` not charge.

<img width="543" alt="Screen Shot 2022-08-26 at 3 23 30 PM" src="https://user-images.githubusercontent.com/43969356/186977053-312dbc06-6fe7-4a57-8249-eb2b43eb37b3.png">
